stats/otel.go

```diff
@@ -61,37 +68,60 @@ func (s *otelStats) Start(ctx context.Context, goFactory GoRoutineFactory) error
 		return fmt.Errorf("failed to create open telemetry resource: %w", err)
 	}
 
-	options := []otel.Option{otel.WithInsecure()} // @TODO: could make this configurable
+	options := []otel.Option{otel.WithInsecure(), otel.WithLogger(s.logger)}
 	if s.otelConfig.tracesEndpoint != "" {
 		options = append(options, otel.WithTracerProvider(
 			s.otelConfig.tracesEndpoint,
 			s.otelConfig.tracingSamplingRate,
 		))
 	}
-	if s.otelConfig.metricsEndpoint != "" {
-		meterProviderOptions := []otel.MeterProviderOption{
-			otel.WithMeterProviderExportsInterval(s.otelConfig.metricsExportInterval),
-		}
-		if len(s.config.defaultHistogramBuckets) > 0 {
+
+	meterProviderOptions := []otel.MeterProviderOption{
+		otel.WithMeterProviderExportsInterval(s.otelConfig.metricsExportInterval),
+	}
+	if len(s.config.defaultHistogramBuckets) > 0 {
+		meterProviderOptions = append(meterProviderOptions,
+			otel.WithDefaultHistogramBucketBoundaries(s.config.defaultHistogramBuckets),
+		)
+	}
+	if len(s.config.histogramBuckets) > 0 {
+		for histogramName, buckets := range s.config.histogramBuckets {
 			meterProviderOptions = append(meterProviderOptions,
-				otel.WithDefaultHistogramBucketBoundaries(s.config.defaultHistogramBuckets),
+				otel.WithHistogramBucketBoundaries(histogramName, defaultMeterName, buckets),
 			)
 		}
-		if len(s.config.histogramBuckets) > 0 {
-			for histogramName, buckets := range s.config.histogramBuckets {
-				meterProviderOptions = append(meterProviderOptions,
-					otel.WithHistogramBucketBoundaries(histogramName, defaultMeterName, buckets),
-				)
-			}
+	}
+	if s.otelConfig.metricsEndpoint != "" {
+		options = append(options, otel.WithMeterProvider(append(meterProviderOptions,
+			otel.WithGRPCMeterProvider(s.otelConfig.metricsEndpoint),
+		)...))
+	} else if s.otelConfig.enablePrometheusExporter {
+		if s.otelConfig.prometheusMetricsPort < 1 || s.otelConfig.prometheusMetricsPort > 65535 {
+			return fmt.Errorf("invalid prometheus metrics port %d", s.otelConfig.prometheusMetricsPort)
 		}
-		options = append(options, otel.WithMeterProvider(s.otelConfig.metricsEndpoint, meterProviderOptions...))
+		options = append(options, otel.WithMeterProvider(append(meterProviderOptions,
+			otel.WithPrometheusExporter(prometheus.DefaultRegisterer),
+		)...))
 	}
 	_, mp, err := s.otelManager.Setup(ctx, res, options...)
 	if err != nil {
 		return fmt.Errorf("failed to setup open telemetry: %w", err)
 	}
 
 	s.meter = mp.Meter(defaultMeterName)
+	if s.otelConfig.enablePrometheusExporter {
+		s.httpServerShutdownComplete = make(chan struct{})
+		s.httpServer = &http.Server{
```

Comment:
I've done it like this because I came across cases where an HTTP server was busy serving other requests and it couldn't serve others that were critical for debugging and monitoring. For that reason I was unable to access the `/debug/pprof` during those times and we (Netdata) therefore decided to split the HTTP server back then. That is why I think we should always have 3 separate HTTP servers:
1. API
2. debug/pprof
3. metrics

I understand that there might be cases like the event schemas where you might want to use the same server. You could still have issues though because you might have middlewares that apply some logic that don't want on debug/pprof or metrics. It would be harder to detect those cases if we merge everything and you would also need conditions on those  middlewares like `if metrics dontDoThis`.

Anyhow, I'll make it configurable but given how cheap it is to have it in another port I personally vote for not having it configurable and always enforce a separate port.

Reply:
@lvrach I made the configuration always mandatory and allowed the port to be `0`.

If you pass a custom registry (now supported [here](https://github.com/rudderlabs/rudder-go-kit/pull/13/commits/a129de81672156464b7c472047ba383ab12ad520)) and pass a `0` port you can now build the handler yourself like:

```go
r := prometheus.NewRegistry()

s := NewStats(c, l, m,
	WithPrometheusRegistry(r, r),
)

// ...

hs := &http.Server{
	Addr: ":9898",
	Handler: promhttp.InstrumentMetricHandler(
		r, promhttp.HandlerFor(r, promhttp.HandlerOpts{
			ErrorLog: &prometheusLogger{l: s.logger},
		}),
	),
}
```

I think this is a good option, let me know your thoughts.

Reply:
Of course a custom registry isn't mandatory, you can still use the default one:

```go
c := config.New()
l := logger.NewFactory(c)
m := metric.NewManager()

s := NewStats(c, l, m)

// ...

hs := &http.Server{
	Addr: ":9898",
	Handler: promhttp.Handler(),
}
```

Reply:
The custom registry approach looks better.